src/Cake/Program.cs

```diff
@@ -1,112 +1,94 @@
-using Cake.Common.Modules;
-using Cake.Composition;
+using Cake.Client.Commands;
+using Cake.Client.Features.Bootstrapping;
+using Cake.Client.Features.Building;
+using Cake.Client.Features.Introspection;
+using Cake.Client.Infrastructure;
+using Cake.Client.Infrastructure.Composition;
+using Cake.Client.Infrastructure.Converters;
 using Cake.Core;
-using Cake.Core.Composition;
-using Cake.Core.Configuration;
 using Cake.Core.Diagnostics;
-using Cake.Core.Modules;
-using Cake.Core.Text;
-using Cake.Modules;
-using Cake.NuGet;
+using Cake.Core.IO;
+using Cake.Core.Reflection;
+using Spectre.Cli;
 
-namespace Cake
+namespace Cake.Client
 {
-    /// <summary>
-    /// The Cake program.
-    /// </summary>
-    public static class Program
+    public sealed class Program
     {
-        /// <summary>
-        /// The application entry point.
-        /// </summary>
-        /// <returns>The application exit code.</returns>
-        public static int Main()
+        private readonly Action<ContainerBuilder> _overrides;
+
+        public Program(Action<ContainerBuilder> overrides = null)
         {
-            ICakeLog log = null;
+            _overrides = overrides;
+        }
 
-            try
-            {
-                // Parse arguments.
-                var args = QuoteAwareStringSplitter
-                    .Split(Environment.CommandLine)
-                    .Skip(1) // Skip executable.
-                    .ToArray();
+        public static async Task<int> Main(string[] args)
+        {
+            var application = new Program();
+            return await application.Run(args);
+        }
 
-                var builder = new ContainerRegistrar();
-                builder.RegisterModule(new CakeModule());
-                builder.RegisterModule(new CoreModule());
-                builder.RegisterModule(new CommonModule());
+        public async Task<int> Run(string[] args, bool propagateExceptions = false)
+        {
+            var registrar = BuildTypeRegistrar();
+
+            var app = new CommandApp<DefaultCommand>(registrar);
+            app.Configure(config =>
+            {
+                config.ValidateExamples();
 
-                // Build the container.
-                using (var container = builder.Build())
+                if (propagateExceptions)
                 {
-                    // Resolve the log.
-                    log = container.Resolve<ICakeLog>();
+                    config.PropagateExceptions();
+                }
+
+                // Top level examples.
+                config.AddExample(new[] { string.Empty });
+                config.AddExample(new[] { "build.cake", "--verbosity", "quiet" });
+                config.AddExample(new[] { "build.cake", "--tree" });
+            });
 
-                    // Parse the options.
-                    var parser = container.Resolve<IArgumentParser>();
-                    var options = parser.Parse(args);
+            return await app.RunAsync(args);
```

Comment:
@patriksvensson maybe `ScriptAssemblyResolver` could be added somewhere around here? Or better yet, before loading any third party modules.